saleor/payment/utils.py

```
@@ -1887,13 +1895,53 @@ def handle_transaction_initialize_session(
         transaction_webhook_response=response_data,
         manager=manager,
     )
-    invalidate_cache_for_stored_payment_methods_if_needed(
-        created_event, source_object, app.identifier
-    )
+
+    if payment_gateway_data.app_identifier != GIFT_CARD_PAYMENT_GATEWAY_ID:
+        invalidate_cache_for_stored_payment_methods_if_needed(
+            created_event,
+            source_object,
+            app.identifier,  # type: ignore[union-attr]
+        )
     data_to_return = response_data.get("data") if response_data else None
     return created_event.transaction, created_event, data_to_return
 
 
+def transaction_initialize_session_with_gift_card_payment_method(
+    transaction_session_data: "TransactionSessionData",
+) -> "TransactionSessionResult":
+    transaction_session_result = TransactionSessionResult(
+        app_identifier=GIFT_CARD_PAYMENT_GATEWAY_ID,
+        response={
+            "result": TransactionEventType.AUTHORIZATION_FAILURE.upper(),
+            "pspReference": str(uuid4()),
+            "amount": transaction_session_data.action.amount,
+        },
+    )
+
+    # Check for existence of an active gift card and validate currency.
+    gift_card_qs = GiftCard.objects.filter(
+        code=transaction_session_data.payment_gateway_data.data["code"],  # type: ignore[call-overload, index]
+        currency=transaction_session_data.action.currency,
+        is_active=True,
+    )
+    if gift_card_qs.count() != 1:
+        return transaction_session_result
```

Shouldn't we add some message or logs why it's failing?

---

After giving it another though I will:
- rely on `code` being unique,
- change the condition to `if gift_card_qs.count() < 1`,
- let the code crash if gift_card_qs.get() tries to return more than one object (which is never going to happen assuming `code` stays unique).